protostar/profiler/contract_profiler.py

```diff
@@ -0,0 +1,319 @@
+            if top.function.id == "starkware.starknet.common.syscalls.call_contract":
+                stack_len = len(instr_callstack)
+                callstacks.append(instr_callstack)
+        return callstacks
+
+    def build_step_samples(
+        self, instructions: List[Instruction], tracer_data: TracerData
+    ) -> List[Sample]:
+        step_samples: List[Sample] = []
+        for trace_entry in tracer_data.trace:
+            callstack = self.get_call_stack(fp=trace_entry.fp, pc=trace_entry.pc)
+            instr_callstack = [
+                self.find_instruction(instructions, pc) for pc in callstack
+            ]
+            step_samples.append((Sample(value=1, callstack=instr_callstack)))
+        return step_samples
+
+    @staticmethod
+    def blame_pc(last_accesses: Dict[Address, Address], hole_address: Address) -> int:
+        """
+        Decides under which pc is placed an intruction which is responsible for the memory hole
+
+        Keyword arguments:
+        last_accesses -- Address -> Pc of instruction which accessed the address last
+        hole_address -- Address of the memory hole
+        """
+        assert hole_address not in last_accesses
+        min_addr_after = math.inf
+        blamed_pc = -1
+        for address, pc in last_accesses.items():
+            if address > hole_address:
+                if address < min_addr_after:
+                    min_addr_after = address
+                    blamed_pc = pc
+        assert blamed_pc > -1
+        return blamed_pc
+
+    @staticmethod
+    def get_not_accessed_addresses(
+        accessed_memory: Set[RelocatableValue],
+        segments: MemorySegmentManager,
+        segment_offsets: Dict[int, int],
+    ) -> Set[Address]:
+        not_accessed_addr: Set[Address] = set()
+        for idx in range(segments.n_segments):
+            size = segments.get_segment_size(segment_index=idx)
+            not_accessed_addr |= {segment_offsets[idx] + i for i in range(size)}
+
+        accessed_offsets_sets: defaultdict[int, Set[int]] = defaultdict(set)
+        for addr in accessed_memory:
+            index, offset = addr.segment_index, addr.offset
+            accessed_offsets_sets[index].add(offset)
+
+        for segment_index, accessed_offsets_set in accessed_offsets_sets.items():
+            for off in accessed_offsets_set:
+                try:
+                    idx: Address = segment_offsets[segment_index] + off
+                    not_accessed_addr.remove(idx)
+                except KeyError:
+                    pass
+        return not_accessed_addr
+
+    def build_memhole_samples(
+        self,
+        instructions: List[Instruction],
+        tracer_data: TracerData,
+        accessed_memory: Set[RelocatableValue],
+        segments: MemorySegmentManager,
+        segment_offsets,
+    ) -> List[Sample]:
+        # Address -> Pc of instruction which accessed the address last
+        accessed_by: Dict[Address, Address] = {}
```

Comment:
I am confused by `Pc of instruction` and `PC` in general. I'd use `InstructionAddress` instead of PC (Program Counter) and renamed the `Address` type to `MemoryAddress`.

Reply:
> MemoryAddress
I do not agree, it is unnecessary wordy

>InstructionAddress
I can consider this in few places

Reply:
Gave it an afterthough: instruction address is a bit better, but pc is used as an internal convention in cairo compiler/vm and i prefer to be consistent as this code touches those parts heavily